src/decoding.rs

```diff
@@ -286,3 +287,58 @@ pub fn decode_header(token: &str) -> Result<Header> {
     let (_, header) = expect_two!(message.rsplitn(2, '.'));
     Header::from_encoded(header)
 }
+
+/// Verify signature of a JWS, and return the header object
+///
+/// If the token or its signature is invalid, it will return an error.
+fn verify_jws_signature<T>(
+    jws: &Jws<T>,
+    key: &DecodingKey,
+    validation: &Validation,
+) -> Result<Header> {
+    if validation.validate_signature && validation.algorithms.is_empty() {
+        return Err(new_error(ErrorKind::MissingAlgorithm));
+    }
+
+    if validation.validate_signature {
+        for alg in &validation.algorithms {
+            if key.family != alg.family() {
+                return Err(new_error(ErrorKind::InvalidAlgorithm));
+            }
+        }
+    }
+
+    let header = Header::from_encoded(&jws.protected)?;
+
+    if validation.validate_signature && !validation.algorithms.contains(&header.alg) {
+        return Err(new_error(ErrorKind::InvalidAlgorithm));
+    }
+
+    let message = [jws.protected.as_str(), jws.payload.as_str()].join(".");
+
+    if validation.validate_signature
+        && !verify(&jws.signature, message.as_bytes(), key, header.alg)?
+    {
+        return Err(new_error(ErrorKind::InvalidSignature));
+    }
+
+    Ok(header)
```

Comment:
We can probably macro it out to avoid duplicating the code from `verify_signature`

Reply:
I tried factoring out a function, let me know if this works for you.